### PR TITLE
Csv copy manager

### PR DESCRIPTION
**Pull request by dmitriy-sych**
New copy manager implementation
Examples of usage you can find in CopyManagerImplTest )

### CODE REVIEW

File: src/main/java/cc/blynk/clickhouse/ClickHouseConnection.java [hunk 1]
```diff
@@ -15,5 +16,7 @@
     @Override
     ClickHouseStatement createStatement(int resultSetType, int resultSetConcurrency) throws SQLException;
 
+    CopyManager createCopyManager();
```
Comment (doom369) [inline src/main/java/cc/blynk/clickhouse/ClickHouseConnection.java]: I think it is better to pass Connection into the `CopyManager`. Or provide constructor for `CopyManager` or even factory method (that is actually much better for public library). Something like:

private final CopyManger copyManager = CopyManager.create(DataSource ds);

```
try (copyManager.copyOut(query, out)) {
}
```

File: src/main/java/cc/blynk/clickhouse/http/DefaultHttpConnector.java [hunk 2]
```diff
@@ -119,24 +128,67 @@ private InputStream sendPostRequest(byte[] sqlBytes, List<byte[]> batches, HttpU
...
+    }
+
+    private byte[] getSqlBytes(String sql) {
+        if (!sql.endsWith("\n")) {
+            sql += "\n";
+        }
+        return sql.getBytes(UTF_8);
+    }
+
+    private void sendPostRequest(InputStream from, OutputStream to, HttpURLConnection connection)
```
Comment (doom369) [inline src/main/java/cc/blynk/clickhouse/http/DefaultHttpConnector.java]: I propose to divide this method into 2 with `OutputStream to` and without.

File: src/test/java/cc/blynk/clickhouse/integration/BatchInsertsTest.java [hunk 3]
```diff
@@ -295,4 +280,18 @@ public void testNullParameters() throws SQLException {
         st.addBatch();
     }
 
+    @BeforeTest
+    public void setUp() throws Exception {
```
Comment (doom369) [inline src/test/java/cc/blynk/clickhouse/integration/BatchInsertsTest.java]: please revert. If you want to restructure code it is better to send separate PR.

File: src/main/java/cc/blynk/clickhouse/copy/CopyManagerImpl.java [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+import java.io.BufferedInputStream;
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.io.Reader;
+import java.io.Writer;
+import java.sql.SQLException;
+
+class CopyManagerImpl implements CopyManager {
+
+    private final ClickHouseConnection connection;
```
Comment (doom369) [inline src/main/java/cc/blynk/clickhouse/copy/CopyManagerImpl.java]: I think holding the connection within `CopyManagerImpl` is wrong. Connection usage should be as short as possible and should be within method scope. And `CopyManagerImpl` should be a singleton and threadsafe. End user should not create `CopyManager` every time he need to execute the query. In real use cases connections are managed by `ConnectionsPool` that hold the references. So it is better to pass the reference to the DataSoruce which will get the connection for method scope.

File: src/main/java/cc/blynk/clickhouse/copy/ReaderInputStream.java [hunk 5]
```diff
@@ -0,0 +1,19 @@
+package cc.blynk.clickhouse.copy;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.Reader;
+
+public class ReaderInputStream extends InputStream {
```
Comment (doom369) [inline src/main/java/cc/blynk/clickhouse/copy/ReaderInputStream.java]: package private

File: src/main/java/cc/blynk/clickhouse/copy/WriterOutputStream.java [hunk 6]
```diff
@@ -0,0 +1,19 @@
+package cc.blynk.clickhouse.copy;
+
+import java.io.IOException;
+import java.io.OutputStream;
+import java.io.Writer;
+
+public class WriterOutputStream extends OutputStream {
```
Comment (doom369) [inline src/main/java/cc/blynk/clickhouse/copy/WriterOutputStream.java]: package private

File: src/test/java/cc/blynk/clickhouse/integration/BatchInsertsTest.java (hunk 3 above)
Comment (dmitriy-sych) [reply src/test/java/cc/blynk/clickhouse/integration/BatchInsertsTest.java]: done

File: src/main/java/cc/blynk/clickhouse/copy/CopyManager.java [hunk 7]
```diff
@@ -0,0 +1,128 @@
...
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.io.Reader;
+import java.io.Writer;
+import java.sql.SQLException;
+
+public interface CopyManager {
+
+    /**
+     * Insert data from stream to DB. Type of data in stream configured in SQL Query.
```
[the remark this answers is not in the document]
Comment (dmitriy-sych) [reply src/main/java/cc/blynk/clickhouse/copy/CopyManager.java]: done

File: src/main/java/cc/blynk/clickhouse/copy/CopyManager.java [hunk 8]
```diff
@@ -0,0 +1,128 @@
...
+     * For example:
+     * <pre>
+     *     CopyManager copyManager = CopyManagerFactory.createCopyManager(connection);
+     *     InputStream is = ...
+     *     String sql = "INSERT INTO default.my_table FORMAT CSV";
+     *     copyManager.copyIn(sql, is);
+     *
+     *     //Data will be written to stream in CSV format
```
[the remark this answers is not in the document]
Comment (dmitriy-sych) [reply src/main/java/cc/blynk/clickhouse/copy/CopyManager.java]: done

File: src/main/java/cc/blynk/clickhouse/copy/CopyManagerFactory.java [hunk 9]
```diff
@@ -0,0 +1,17 @@
+package cc.blynk.clickhouse.copy;
+
+import cc.blynk.clickhouse.ClickHouseConnection;
+
+public final class CopyManagerFactory {
+
+    private CopyManagerFactory() {
+    }
+
+    public static CopyManager createCopyManager(ClickHouseConnection connection) {
```
[the remark this answers is not in the document]
Comment (dmitriy-sych) [reply src/main/java/cc/blynk/clickhouse/copy/CopyManagerFactory.java]: done.

**Comment by doom369** [comment]
@dmitriy-sych still build failed:

```
[ERROR] Failures: 
[ERROR]   CopyManagerImplTest.copyOutStreamTest:136 expected ["date","date_time","string","int32","float64"
"1989-01-30","2016-08-12 13:21:32","testString",2147483647,42.21
] but found ["date","date_time","string","int32","float64"
"1989-01-29","2016-08-12 16:21:32","testString",2147483647,42.21
]
[ERROR]   CopyManagerImplTest.copyOutWriterTest:147 expected ["date","date_time","string","int32","float64"
"1989-01-30","2016-08-12 13:21:32","testString",2147483647,42.21
] but found ["date","date_time","string","int32","float64"
"1989-01-29","2016-08-12 16:21:32","testString",2147483647,42.21
]
[ERROR]   CsvCopyManagerImplTest.copyOutStreamTest:64 expected ["1989-01-30","2016-08-12 13:21:32","testString",2147483647,42.21
] but found ["1989-01-29","2016-08-12 16:21:32","testString",2147483647,42.21
]
[ERROR]   ClickHousePreparedStatementRealTablesTest.testBatchInsertOf1kkOfReocordsWithoutCompression:111 » OutOfMemory
```

File: src/main/java/cc/blynk/clickhouse/http/DefaultHttpConnector.java (hunk 2 above)
Comment (dmitriy-sych) [reply src/main/java/cc/blynk/clickhouse/http/DefaultHttpConnector.java]: done

File: src/main/java/cc/blynk/clickhouse/ClickHouseConnection.java (hunk 1 above)
Comment (dmitriy-sych) [reply src/main/java/cc/blynk/clickhouse/ClickHouseConnection.java]: factory added
